src/bio/bio_xstream.c

```diff
@@ -88,6 +91,228 @@ struct bio_nvme_data {
+{
+	int	sc; /* status_code */
+	int	sct; /* status_code_type */
+
+	D_PRINT("in get_spdk_log_page_completion\n");
+	if (!success)
+		D_PRINT("SPDK bdev I/O failed\n");
+	else
+		D_PRINT("SPDK bdev I/O succeeded\n");
+	/* additional error information */
+	spdk_bdev_io_get_nvme_status(bdev_io, &sct, &sc);
+	D_PRINT("NVMe status code: %d, type: %d\n", sc, sct);
+	/*
+	 * status code = INTERNAL_DEVICE_ERROR = 6
+	 * type: 0 = SPDK_NVME_SCT_GENERIC
+	 */
+}
+
+static void
+dprint_uint128_hex(uint64_t *v)
+{
+	unsigned long long lo = v[0], hi = v[1];
+	if (hi) {
+		D_PRINT("0x%llX%016llX", hi, lo);
+	} else {
+		D_PRINT("0x%llX", lo);
+	}
+}
+
+static void
+dprint_uint128_dec(uint64_t *v)
+{
+	unsigned long long lo = v[0], hi = v[1];
+	if (hi) {
+		/* can't handle large (>64-bit) decimal values */
+		dprint_uint128_hex(v);
+	} else {
+		D_PRINT("%llu", (unsigned long long)lo);
+	}
+}
+
+/*
+ * Query the SPDK health & error information log pages to keep device health
+ * in-memory state up-to-date and catch any potential errors early.
+ * TODO Decide on cadence for querying info, currently every 60 seconds.
+ */
+static void
+query_spdk_health_stats(struct bio_xs_context *ctxt, uint64_t now)
+{
+/*	const struct spdk_nvme_ctrlr_data		*cdata; */
+/*	struct spdk_nvme_error_information_entry	*error_entry; */
+/*	uint32_t					 i; */
+	struct spdk_bdev				*bdev = NULL;
+	struct spdk_io_channel				*channel = NULL;
+	struct spdk_nvme_cmd				*cmd = NULL;
+	uint64_t					 offset = 0;
+	uint32_t					 payload_size;
+	uint32_t					 numd, numdl, numdu, lpol, lpou;
+	int						 rc;
+
+	 if (ctxt->bxc_health_stat_age + DAOS_SPDK_STATS_PERIOD >= now)
+		return;
+
+	/* used example in spdk_nvme_ctrlr_cmd_get_log_page() to set cmd parameters */
+	payload_size = sizeof(health_page);
+	if (payload_size == 0)
+		return;
+
+	numd = payload_size / sizeof(uint32_t) - 1u;
+	numdl = numd & 0xFFFFu;
+	numdu = (numd >> 16) & 0xFFFFu;
+	lpol = (uint32_t)offset;
+	lpou = (uint32_t)(offset >> 32);
+
+	D_ALLOC_PTR(cmd);
+	if (cmd == NULL)
+		return;
+	memset(cmd, 0, sizeof(struct spdk_nvme_cmd));
+	cmd->opc = SPDK_NVME_OPC_GET_LOG_PAGE;
+	cmd->nsid = SPDK_NVME_GLOBAL_NS_TAG;
+	cmd->cdw10 = numdl << 16;
+	cmd->cdw10 |= SPDK_NVME_LOG_HEALTH_INFORMATION;
+	cmd->cdw11 = numdu;
+	cmd->cdw12 = lpol;
+	cmd->cdw13 = lpou;
+
+	if (ctxt->bxc_desc != NULL) {
+		channel = spdk_bdev_get_io_channel(ctxt->bxc_desc);
+		D_ASSERT(channel != NULL);
+
+		bdev = spdk_bdev_desc_get_bdev(ctxt->bxc_desc);
+
+		/* example from spdk_rpc_apply_nvme_firmware() */
+		/* check whether the block device supports admin I/O before passthru */
+		if (!spdk_bdev_io_type_supported(bdev, SPDK_BDEV_IO_TYPE_NVME_ADMIN)) {
+			D_ERROR("SPDK_BDEV_IO_TYPE_NVME_ADMIN not supported\n");
+			goto cleanup;
+		}
+
+		/*
+		 * Use spdk_bdev_nvme_admin_passthru() to submit an NVMe admin
+		 * command to bdev.
+		 */
+		rc = spdk_bdev_nvme_admin_passthru(ctxt->bxc_desc,
+						   channel,
+						   cmd,
+						   &health_page,
```

Comment:
The buffer passed should be DMA safe buffer, you'd allocate it by spdk_dma_malloc().